README
NAME
    Module::Install::CPANfile - Include cpanfile

SYNOPSIS
      # cpanfile
      requires 'Plack', 0.9;
      on test => sub {
          requires 'Test::Warn';
      };

      # Makefile.PL
      use inc::Module::Install;
      cpanfile;
      WriteAll;

DESCRIPTION
    Module::Install::CPANfile is a plugin for Module::Install to include
    dependencies from cpanfile into meta files. "MYMETA" files are always
    merged with cpanfile, as well as "META" files when you run "Makefile.PL"
    as an author mode (as in preparation for "make dist").

WHY?
    If you have read cpanfile and cpanfile-faq you might wonder why this
    plugin is necessary, since *cpanfile* is meant to be used by Perl
    applications and not distributions.

    Well, for a couple reasons.

  Co-exist as a Github project and CPAN distribution
    One particular use case is a script or web application that you develop
    on github, and you start with writing dependencies in *cpanfile*. Other
    users or developers pull the code from git, use carton to bundle
    dependencies. So far, so good.

    One day you decide to release the script or application to CPAN as a
    CPAN distribtuion, and that means you have to duplicate *cpanfile* into
    *Makefile.PL* or *Build.PL*, and that's not DRY! This plugin will allow
    you to have just one *cpanfile* and reuse it from *Makefile.PL* for CPAN
    client use.

  CPAN Meta Spec 2.0 support.
    The other, kind of unfortunate reason is that as of this writing,
    Module::Install and ExtUtils::MakeMaker don't have a complete support
    for expressing CPAN::Meta::Spec version 2.0 vocabularies.

    There are many of known issues that the metadata you specify with
    Module::Install are somehow discarded or down converted, and eventually
    lost in *MYMETA* files. The examples of those issues are that test
    dependencies gets merged with build dependencies, and that *recommends*
    is not saved into MYMETA files.

    This plugin lets you write prerequisites in *cpanfile* which is more
    powerful and has a complete support for CPAN::Meta::Spec 2.0 and makes
    your application forward compatible.

AUTHOR
    Tatsuhiko Miyagawa <miyagawa@bulknews.net>

COPYRIGHT
    Copyright 2012- Tatsuhiko Miyagawa

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    cpanfile Module::CPANfile Module::Install



Changes
Revision history for Perl extension Module::Install::CPANfile

0.11  Thu Jan 31 15:06:30 PST 2013
        - Experimental: Enable dynamic_config 0 by default
        - Increase the dependency for and include the latest Module::CPANfile

0.10  Wed Jan 30 19:48:04 PST 2013
        - Major update to cpanfile command: now it actually updates MYMETA
          files upon installation, bypassing M::I and EUMM conversion
        - Checks all requirements in author mode using CPAN::Meta::Check

0.04  Fri Apr 13 18:13:45 CST 2012
        - Fixed dependencies
        - Added xt tests (xaicron)

0.03  Thu Apr 12 20:51:21 JST 2012
        - Fix version range to a minimum version requirement so that Module::Install can understand (xaicron)

0.02  Thu Apr 12 20:25:44 JST 2012
        - Fix runtime requires

0.01  Thu Apr 12 19:46:19 2012
        - original version


xt/Util.pm
package xt::Util;

use strict;
use warnings;
use Test::Builder;
use File::Temp qw(tempdir);
use File::Path qw(mkpath);
use File::Basename qw(dirname);
use Cwd qw(getcwd);
use Config qw(%Config);
use YAML::Tiny;
use Capture::Tiny qw(capture_merged);
use base qw(Exporter);

use constant DMAKE => $^O eq 'MSWin32' && $Config{make} =~ /dmake(?:\.exe)?$/i;
use constant NMAKE => $^O eq 'MSWin32' && $Config{make} =~ /nmake(?:\.exe)?$/i;
use constant MAKE  => $Config{make} || 'make';

our @EXPORT = qw/make_meta_data unpack_tree build run_make DMAKE NMAKE/;

sub make_meta_data {
    my $fh = shift;

    local $ENV{PERL5LIB} = join $Config{path_sep},
        map { File::Spec->rel2abs($_) } @INC;

    my $cwd    = getcwd;
    my $tmpdir = tempdir CLEANUP => 1;
    chdir $tmpdir or die $!;

    unpack_tree($fh);

    my $yaml = eval {
        build($cwd);
        YAML::Tiny->read('META.yml');
    };
    chdir $cwd or die $!;

    die $@ if $@;

    return $yaml->[0];
}

sub build {
    my $distdir = shift;
    die "Makefile.PL not found" unless -f 'Makefile.PL';
    my $tb = Test::Builder->new;
    $tb->note( run_cmd(qq{$^X Makefile.PL}) );
    run_make();
}

sub run_cmd {
    my ($cmd) = @_;
    my $result = capture_merged {
        system $cmd;
    };
    die "`$cmd` failed ($result)" if $?;
    return $result;
}

sub run_make {
    run_cmd(join ' ', MAKE, @_);
}

sub _parse_data {
    my $fh = shift;
    my ($data, $path);
    while (<$fh>) {
        if (/^\@\@/) {
            ($path) = $_ =~ /^\@\@ (.*)/;
            next;
        }
        $data->{$path} .= $_;
    }
    close $fh;
    return $data;
}

sub unpack_tree {
    my $data = _parse_data(shift);

    for my $path (keys %$data) {
        my $dir = dirname($path);
        unless (-e $dir) {
            mkpath($dir) or die "Cannot mkpath '$dir': $!";
        }

        my $content = $data->{$path};
        open my $out, '>', $path or die "Cannot open '$path' for writing: $!";
        print $out $content;
        close $out;
    }
}

sub _regex {
    my @target = @_;
    my $regex = join '|', @target;
    return qr/$regex/;
}

1;
__END__


lib/Module/Install/CPANfile.pm
package Module::Install::CPANfile;

use strict;
use 5.008_001;
our $VERSION = '0.11';

use Module::CPANfile;
use base qw(Module::Install::Base);

sub merge_meta_with_cpanfile {
    my $self = shift;

    require CPAN::Meta;

    my $file = Module::CPANfile->load;

    if ($self->is_admin) {
        # force generate META.json
        CPAN::Meta->load_file('META.yml')->save('META.json');

        print "Regenerate META.json and META.yml using cpanfile\n";
        $file->merge_meta('META.yml');
        $file->merge_meta('META.json');
    }

    for my $metafile (grep -e, qw(MYMETA.yml MYMETA.json)) {
        print "Merging cpanfile prereqs to $metafile\n";
        $file->merge_meta($metafile);
    }
}

sub cpanfile {
    my($self, %options) = @_;

    $self->include("Module::CPANfile");
    $self->configure_requires("CPAN::Meta");

    $self->dynamic_config(0) unless $options{dynamic};

    my $write_all = \&::WriteAll;

    *main::WriteAll = sub {
        $write_all->(@_);
        $self->merge_meta_with_cpanfile;
    };

    $self->include("Module::CPANfile");
    $self->configure_requires("CPAN::Meta");

    if ($self->is_admin) {
        if (eval { require CPAN::Meta::Check; 1 }) {
            my $prereqs = Module::CPANfile->load->prereqs;
            my @err = CPAN::Meta::Check::verify_dependencies($prereqs, [qw/runtime build test develop/], 'requires');
            for (@err) {
                warn "Warning: $_\n";
            }
        } else {
            warn "CPAN::Meta::Check is not installed. Skipping dependencies check for the author.\n";
        }
    }
}

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

Module::Install::CPANfile - Include cpanfile

=head1 SYNOPSIS

  # cpanfile
  requires 'Plack', 0.9;
  on test => sub {
      requires 'Test::Warn';
  };

  # Makefile.PL
  use inc::Module::Install;
  cpanfile;
  WriteAll;

=head1 DESCRIPTION

Module::Install::CPANfile is a plugin for Module::Install to include
dependencies from L<cpanfile> into meta files. C<MYMETA> files are
always merged with cpanfile, as well as C<META> files when you run
C<Makefile.PL> as an author mode (as in preparation for C<make dist>).

=head1 WHY?

If you have read L<cpanfile> and L<cpanfile-faq> you might wonder why
this plugin is necessary, since I<cpanfile> is meant to be used by
Perl applications and not distributions.

Well, for a couple reasons.

=head2 Co-exist as a Github project and CPAN distribution

One particular use case is a script or web application that you
develop on github, and you start with writing dependencies in
I<cpanfile>. Other users or developers pull the code from git, use
L<carton> to bundle dependencies. So far, so good.

One day you decide to release the script or application to CPAN as a
CPAN distribtuion, and that means you have to duplicate I<cpanfile>
into I<Makefile.PL> or I<Build.PL>, and that's not DRY! This plugin
will allow you to have just one I<cpanfile> and reuse it from
I<Makefile.PL> for CPAN client use.

=head2 CPAN Meta Spec 2.0 support.

The other, kind of unfortunate reason is that as of this writing,
L<Module::Install> and L<ExtUtils::MakeMaker> don't have a complete
support for expressing L<CPAN::Meta::Spec> version 2.0
vocabularies.

There are many of known issues that the metadata you specify with
L<Module::Install> are somehow discarded or down converted, and
eventually lost in I<MYMETA> files. The examples of those issues are
that test dependencies gets merged with build dependencies, and that
I<recommends> is not saved into MYMETA files.

This plugin lets you write prerequisites in I<cpanfile> which is more
powerful and has a complete support for L<CPAN::Meta::Spec> 2.0 and
makes your application forward compatible.

=head1 AUTHOR

Tatsuhiko Miyagawa E<lt>miyagawa@bulknews.netE<gt>

=head1 COPYRIGHT

Copyright 2012- Tatsuhiko Miyagawa

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<cpanfile> L<Module::CPANfile> L<Module::Install>

=cut
